### DISCUSSION TITLE
Close rofi after selection

### DISCUSSION THREAD

**Discussion by Gijom** [davatorium/rofi#1850] [Q&A] [answered]
I am trying to write a mode script to handle UNIX [pass](https://www.passwordstore.org/) manager. I know several solutions already exist but most of them do not have the functionality I need or are implemented with rofi-dmenu. Notice that when using [pass](https://www.passwordstore.org/), a pinentry graphical menu pops to prompt for the GPG password.

The problem with the rofi-script mode is that the rofi menu does not disappear when the pinentry menu pops, which makes it impossible to type the password (rofi window is above the pinentry window).
An example of such a script is available [here](https://github.com/Gijom/rofi-pass-modi) in the master branch.

As mentioned before using a full rofi-dmenu mode would not allow integration with other rofi modi which I aim for.

The solution I propose is an intermediate hack which aims at providing the best of both worlds:
- it starts as a rofi-script to list password files (which does not need a pinentry)
- on the second (same) rofi-script call it uses coproc to start the script in rofi-dmenu mode
The script/dmenu is available in the [dmenu branch](https://github.com/Gijom/rofi-pass-modi/tree/dmenu).

The main disadvantage is that it forbids switching to another mode once a password is selected. Although once a users have selected a password it is unlikely that they wish to switch mode.

My questions are:
- do you think this solution is appropriate or does it break rofi "design rules" ?
- would there be a way to implement it all as a rofi-script (i.e. no rofi-dmenu), for instance by having the rofi window disappearing during the script processing ?

rofi version: rofi 1.75

**Comment by Gijom** [comment]
After a bit more diffing I noticed that this is related to [this discussion](https://github.com/davatorium/rofi/discussions/1486).
However my problem is not in dmenu mode given that my script asks for a password before `rofi -dmenu` is called.

**Reply by DaveDavenport** [reply, marked as answer]
I am not sure I completely follow this story. 
So Ill try to answer the few things I understood:
1.) Normal rofi grabs all user input, another application popping up while rofi is still running won't be able to get input, this is also why you should not run rofi in rofi. This is by design. This behavior can be changed by running rofi in `-normal-window` mode, then you can switch away to other applications while running. In this mode I guess pinentry  will be able to get input for pass entry.
2.) One of the reasons people use script mode is so the window does not disappears.   In theory we could add a flag to tell rofi to hides it window.  This is currently not supported and I regrettable do not have time to look at this.
3.) I don't think there are real 'design rules'. I might not expect and take it into account in changes and therefor  cannot guarantee it will work identical in future versions.

(sorry reply to first post mostly)

**Comment by Gijom** [comment]
Thanks for your feedback.

1) I noticed the `-normal-window` option but than the rofi-script depends on it which I find a bit weird. Is there a way to set some options from scripts or to check their value, for instance to output a warning message.

2) if you implement it one day I think rofi should also stop grabing user input when the window is not active.

3) duly noted, I think my solution is probably flexible and simple enough that it should hopefuly be resilient to changes.

**Reply by DaveDavenport** [reply]
> if you implement it one day I think rofi should also stop grabing user input when the window is not active.

This is normal-window mode. 
In it normal operation mode (how it configures the window to be unmanaged and on top of everything ) this would not work.  So to make it work, you get '-normal-window' mode.

**Reply by Gijom** [reply]
Ok so it seems normal window mode is one (the?) way to go. Would there be a way to force this option in a script or warn the user that it should be enabled for the script to work properly ?
